File: backend/app/api/sync.py
"""Trigger a sync and poll its progress."""

from __future__ import annotations

from fastapi import APIRouter, Query

from ..schemas.sync import SyncStatus
from ..services import sync_manager

router = APIRouter(prefix="/api/sync", tags=["sync"])


def _to_status(s: sync_manager.SyncState) -> SyncStatus:
    return SyncStatus(
        running=s.running,
        fetched=s.fetched,
        total=s.total,
        done=s.done,
        error=s.error,
        hint=s.hint,
        kind=s.kind,
        detail=s.detail,
    )


@router.post("", response_model=SyncStatus)
async def trigger_sync(full: bool = Query(False)) -> SyncStatus:
    await sync_manager.start_sync(full=full)
    return _to_status(sync_manager.get_state())


@router.get("/status", response_model=SyncStatus)
async def sync_status() -> SyncStatus:
    return _to_status(sync_manager.get_state())

File: frontend/src/api/sync.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { api } from './client';

export interface SyncStatus {
  running: boolean;
  fetched: number;
  total: number;
  done: boolean;
  error: string | null;
  hint: string | null;
  kind: string | null;
}

export function useSyncStatus() {
  return useQuery({
    queryKey: ['sync-status'],
    queryFn: () => api.get<SyncStatus>('/api/sync/status'),
    // Poll quickly while a sync is running; stop otherwise.
    refetchInterval: (q) => (q.state.data?.running ? 800 : false),
  });
}

export function useTriggerSync() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (full: boolean = false) => api.post<SyncStatus>(`/api/sync?full=${full}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['sync-status'] });
    },
  });
}
